cli_tools/gce_vm_image_import/main.go

```diff
@@ -18,119 +18,117 @@ package main
-	labels               = flag.String("labels", "", "List of label KEY=VALUE pairs to add. Keys must start with a lowercase character and contain only hyphens (-), underscores (_), lowercase characters, and numbers. Values must contain only hyphens (-), underscores (_), lowercase characters, and numbers.")
-	storageLocation      = flag.String("storage_location", "", "Location for the imported image which can be any GCS location. If the location parameter is not included, images are created in the multi-region associated with the source disk, image, snapshot or GCS bucket.")
-	uefiCompatible       = flag.Bool("uefi_compatible", false, "Enables UEFI booting, which is an alternative system boot method. Most public images use the GRUB bootloader as their primary boot method.")
-	sysprepWindows       = flag.Bool("sysprep_windows", false, "Whether to generalize image using Windows Sysprep. Only applicable to Windows.")
-)
+func main() {
+	log.SetPrefix("[import-image] ")
+	ctx := context.Background()
 
-func importEntry() (service.Loggable, error) {
+	importArgs, err := parseAndPopulateAllArgs(ctx)
+	if err != nil {
+		terminate(importArgs, err)
+	}
 
-	ctx := context.Background()
-	metadataGCE := &compute.MetadataGCE{}
+	importerClosure := func() (service.Loggable, error) {
+		wf, e := importer.NewImporter(importArgs.Environment, importArgs.Image, importArgs.Translation).Run(ctx)
+		return service.NewLoggableFromWorkflow(wf), e
+	}
+
+	project := importArgs.Environment.Project
+	if err := service.RunWithServerLogging(
+		service.ImageImportAction, initLoggingParams(importArgs), &project, importerClosure); err != nil {
+		os.Exit(1)
+	}
+}
+
+func parseAndPopulateAllArgs(ctx context.Context) (args.ImporterArguments, error) {
+	// 1. Parse auth flags, which are required to initialize the API clients
+	// that are used for subsequent validation and population.
+	fs := flag.NewFlagSet("auth-flags", flag.ContinueOnError)
+	oauth := flag.String("oauth", "", "Path to oauth json file.")
+	ce := flag.String("compute_endpoint_override", "", "API endpoint to override default.")
+	// Don't write parse errors to stdout, instead propagate them via an
+	// exception since we use flag.ContinueOnError.
+	fs.SetOutput(ioutil.Discard)
+	// Ignoring parse errors here since FlagSet.Parse reports that there are extra
+	// flags (such as client_id) that are not defined. That's okay: we just want the
+	// authentication flags now, and we'll re-parse everything next.
+	_ = fs.Parse(os.Args[1:])
 
+	// 2. Setup dependencies.
 	storageClient, err := storage.NewStorageClient(ctx, logging.NewDefaultLogger(), *oauth)
 	if err != nil {
-		return nil, err
+		terminate(args.ImporterArguments{}, err)
 	}
-
+	sourceFactory := importer.NewSourceFactory(storageClient)
 	computeClient, err := param.CreateComputeClient(&ctx, *oauth, *ce)
 	if err != nil {
-		return nil, err
+		return args.ImporterArguments{}, err
```

Comment:
Why is this handled differently from the error above? Wouldn't this result in no args being logged since this object is initialized without being populated with flag values?

Reply:
Good point. I separated parsing, dependency construction, and validation so that we can capture the intermediate values if there is an intermediate failure.